src/ocean/util/container/btree/BTreeMap.d

```diff
@@ -0,0 +1,590 @@
+
+        Recursive inorder iteration over values only. Note that, in case the
+        tree is changed during the iteration, iteration will halt.
+
+        Params:
+            dg = opApply's delegate
+
+        Returns:
+            return value of dg
+
+    ***********************************************************************/
+
+    public int opApply (int delegate (ref ValueType) dg)
+    {
+        return this.impl.inorder(dg);
+    }
+
+
+    /// Convenience alias for the implementation
+    package alias BTreeMapImplementation!(KeyType, ValueType, tree_degree)
+        Implementation;
+
+    /// Convenience alias for the key type
+    package alias TreeKeyType KeyType;
+
+    /// Convenience alias for the element type
+    package alias TreeValueType ValueType;
+
+    /// Private implementation
+    package Implementation impl;
+}
+
+/*******************************************************************************
+
+    Constructor-like method. Constructs the BTreeMap and initializes it.
+
+    Params:
+        KeyType = type of the key.
+        ValueType = type of the value to store.
+        tree_degree = degree of the tree (refer to the documentation)
+        memManager = memory management strategy to use
+
+    Returns:
+        empty BTreeMap which maps KeyType to ValueType of a given degree
+        which sues memManager allocation strategy.
+
+******************************************************************************/
+
+public BTreeMap!(KeyType, ValueType, tree_degree) makeBTreeMap
+    (KeyType, ValueType, int tree_degree)(IMemManager allocator = mallocMemManager)
+{
+    auto tree = BTreeMap!(KeyType, ValueType, tree_degree).init;
+    tree.initialize(allocator);
+    return tree;
+}
+
+version (UnitTest)
+{
+    import ocean.util.container.btree.BTreeMapRange;
+}
+
+///
+unittest
+{
+    // import ocean.util.container.btree.BTreeMapRange;
+    struct MyVal
+    {
+        int x;
+    }
+
+    auto map = makeBTreeMap!(int, MyVal, 2);
+
+    for (int i = 0; i <= 10; i++)
+    {
+        map.insert(i, MyVal(i*2));
+    }
+
+    // find the element by key
+    bool found;
+    auto val = map.get(5, found);
+    test!("==")(found, true);
+    test!("==")(val.x, 10);
+
+    // remove the element
+    map.remove(10);
+    map.get(10, found);
+    test!("==")(found, false);
+
+    // iterate over using opApply
+    foreach (key, val; map)
+    {
+        test!("==")(key * 2, val.x);
+    }
+
+    // iterate over using range
+    void[] buff;
+    for(auto range = byKeyValue(map, &buff);
+            !range.empty; range.popFront())
+    {
+        test!("==")(range.front.key * 2, range.front.value.x);
+    }
+
+    // Let's check that the memory is ready to be reused
+    test(buff.ptr !is null);
+}
+
+/*
+
+    Unittests. Compile with -debug=BTreeMapSanity to turn on
+    the invariant checks for the tree.
+
+*/
+
+version (UnitTest)
+{
+    import ocean.core.Test;
+}
+
+unittest
+{
```

Comment:
Might be nice to break this down into separate unittests, to make it easier to follow. There's at very least the test for the file map and the tests for the integer map.